drafts/cypress-app-actions-with-angular.md

```diff
@@ -0,0 +1,138 @@
+
+First, we needed to make the desired service accessible on the window object.
+We decided against overusing app actions for the time being.
+The interface, respectively dependency, between Cypress and our Angular app should be kept small.
+As we had a `BackendService` that centrally handled all requests to our API, it was an obvious choice.
+It would be only one connection point while offering a high degree of flexibility.
+In the constructor of the service, we verify whether the app is running inside Cypress and then expose the service.
+This way, it's not accessible in any other situation.
+
+`src/app/.../backend.service.ts`:
+```typescript
+export class BackendService {
+    constructor() {
+        if (window.hasOwnProperty('Cypress')) {
+            window.app4cy = window.app4cy ?? {};
+            window.app4cy.BackendService = this;
+        }
+    }
+
+    public get<T>(url: string): Promise<T> {
+        // ...
+    }
+
+    // ...
+}
+```
+
+# Abstraction through Commands
+
+The action should be easy to use in Cypress, so we created a custom command: `backendRequest`.
+This way, we keep the dependency on the app in one central place and abstract the app action.
+
+`cypress/support/commands.ts`:
+```typescript
+Cypress.Command.add('backendRequest',
+    {prevSubject: false},
+    <M extends keyof BackendService>(method: M, ...args: Parameters<BackendService[M]>) =>
+    cy.log(`Use BackendService for a ${method.toUpperCase()} request to '${args[0]}'`)
+        .window({log: false})
+        .its('app4cy.BackendService', {log: false})
+        // @ts-ignore (TS compiler sadly struggles with the relationship between the types of "method" and "args")
+        .then(async (service: BackendService) => service[method](..args));
+);
+```
+
+That also allowed us to specify the types of parameters and return values.
+
+`cypress/support/commands.ts`:
+```typescript
+declare global {
+    namespace Cypress {
+        interface Chainable<Subject> {
+            backendRequest<M extends keyof BackendService>(
+                method: M,
+                ...args: Parameters<BackendService[M]>
+            ): Chainable<ReturnType<BackendService[M]>>;
+        }
+    }
+}
+```
+
+We then went even one step further.
+Repeated actions, such as creating a specific item, were put into a utility function.
+One could argue this was unnecessary, but depending on your needs it might prove to be helpful.
+The actual test looks more readable when calling `createNewItem("Just do it")` than understanding what that request is doing.
+
+`cypress/support/utils.ts`:
+```typescript
+export function createNewItem(itemName: string): void {
+    const urlPath = 'items';
+    cy.backendRequest('post', urlPath, itemName);
+}
+```
+
+# Conclusion
+
+Once we set this app action up, the affected tests greatly improved stability, readability, and speed.
+Using app actions halved the duration of some tests.
+Unexpected side effects and resolved most unwanted dependencies.
```

Comment:
There is something missing in this sentence. What happend to the unexpected side effects? Maybe even combine it with the sentence before it

Reply:
Removed "side effects" part, it's unclear and a bit redundant.